Add Signatures and Signing Keys

cohix opened the pull request:
This PR adds everything needed to sign Invoices, along with some convenience methods for working with a `keyring.toml` file, generating keys, etc.

This has been tested against the bindle server on the main branch as of the day this was opened (`3de704c8e66f50bb956938c6d63915149cf13ca1`)

cohix committed:
use default HTTP client when tlsConfig is nil

cohix committed:
add initial implementation of Invoice Signature

tweak how cleartext is generated

removed v from version, added debug log line

cohix committed:
remove timestamp from cleartext

remove debug log line

cohix committed:
add Keyring and SignatureKey

cohix committed:
add convenience functions for keyring/signaturekey

better handling of keyring

cohix committed:
base64 encode label signature

cohix committed:
update keyringFilepath logic

update comment

update comment

address review feedback and fix test server

ghost commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/deislabs/go-bindle?pullRequest=4) <br/>All CLA requirements met.

cohix commented:
@thomastaylor312 I'd appreciate a review :)

thomastaylor312 commented on types/signature.go:
```diff
@@ -0,0 +1,72 @@
+// 098fa798779ac88094b6d54a3f5cdba41fe5a901
+// 5b992e90b71d5fadab3cd3777230ef370df75f5b
+
+// NOTE: the spec (https://github.com/deislabs/bindle/blob/main/docs/signing-spec.md#signing-on-the-invoice)
```
Good catch on this.

thomastaylor312 commented on keyring/keyring.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+func keyringFilepath() string {
+	base := filepath.Join("$HOME", ".bindle")
```
Will this default work on windows?

cohix replied:
No, but I'm adding an additional option (real homeDir detection) so that there are at least two things that need to fail in order for this fallback to be chosen :)

thomastaylor312 commented on keyring/keyring.go:
```diff
@@ -0,0 +1,110 @@
+
+// GenerateSigningKey generates a keypair for signing Bindle invoices
+// The return types are the public key (wrapped in a SignatureKey), the private key, and any error
+func GenerateSignatureKey(author, role string) (*types.SignatureKey, []byte, error) {
```
It would probably be good to validate the given role to make sure it is one of the supported roles. Right now someone could create and use a key that has an unsupported role

thomastaylor312 commented on types/signature.go:
```diff
@@ -0,0 +1,72 @@
+
+	cleartext := i.generateCleartext("creator", timestamp)
+
+	sig := ed25519.Sign(privKey, []byte(cleartext))
```
This should have a check to make sure the incoming signatureKey has the `creator` role in its list

cohix replied:
Is this check still required if I change the method to accept any author/role combo?

cohix replied:
Answered my own question, will just use the `role` on the SignatureKey, duh :)

cohix replied:
Or at least check to see that the given role is included in the key's list of roles :P

thomastaylor312 commented on types/signature.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	return strings.Join(cleartextParts, "\n")
+}
```
Can you also add a generic `sign` method that can be used to sign for any user, role, and key? Right now I can only add creator signatures

thomastaylor312 reviewed (changes requested):
This is looking great! Thank you for doing this. In addition to the comments, can you also add a test or two to validate that things are signed properly? Bonus points if you want to add a full integration test to make sure the server validates something you've signed (but definitely not required)

thomastaylor312 commented:
Note, the tests are failing because we haven't updated this library in a while. Just need to add a flag to the integration tests. If you have any problems, I can open up a quick PR to fix it. Just let me know!

cohix commented:
@thomastaylor312 I've addressed your comments and (pretty sure) fixed the integration tests (works locally)... adding some tests for the signing keys now.

cohix committed:
added NewParcel function, invoice.VerifySignatures method, and tests for signatures

cohix commented:
@thomastaylor312 added some tests and more convenience funcs, I think this is good for another review :)

radu-matei commented:
I think CI is failing because the Go version in `go.mod` is 1.15, and the `os` package was introduced in 1.16:

```
# github.com/deislabs/go-bindle/keyring
keyring/keyring.go:45:23: undefined: os.ReadFile
keyring/keyring.go:78:12: undefined: os.WriteFile
keyring/keyring.go:89:12: undefined: os.WriteFile
keyring/keyring.go:98:19: undefined: os.ReadFile
```

Bumping the version in `go.mod` should use the correct version (the cached Go [versions available in the runner are Go 1.15.15, 1.16.13, 1.17.6](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#go)).

thomastaylor312 commented on keyring/keyring.go:
```diff
@@ -0,0 +1,118 @@
+	"github.com/pelletier/go-toml"
+)
+
+// GenerateSigningKey generates a keypair for signing Bindle invoices
```
Looks like you'll need to update the function name here so it counts as the doc comment

thomastaylor312 commented on tests/client_integration_test.go:
```diff
@@ -20,11 +21,15 @@ import (
 )
 
+const testAuthor = `Testy McTestface <"testy@test.face">`
+const testAuthor2 = `Elon Testla <"elon@testla.com">`
```
🤣

thomastaylor312 commented on tests/client_integration_test.go:
```diff
@@ -73,11 +78,12 @@ func newTestController(t *testing.T) testController {
-		"-i",
-		address,
-		"-c", cert, "-k", key)
+		"-d", tempdir,
```
The ordering of these flags forms an...unintentional word choice I didn't notice until now. Should probably switch it 😅

cohix replied:
LOL done.

thomastaylor312 commented on types/signature.go:
```diff
@@ -0,0 +1,174 @@
+// includes the `at` value in the cleartext, but the server does not, so this client does not either.
+// Issue: https://github.com/deislabs/bindle/issues/284
+
+// GenerateSignature generates a signature for the privided role and author,
```
Nit (spelling): "provided"

thomastaylor312 commented on types/signature.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	// then validate each of the invoice's signatures using the keys we have available
+	for _, s := range i.Signature {
```
This is basically doing the equivalent of `ExhaustiveVerification`. I think that is fine for a first pass at this, but it quickly could become a bit of a nightmare as you have to validate (and have a public key for) every single signature on an invoice. If we merge as is, we should open an issue to implement the other verification strategies

cohix replied:
You make a good point... I've tweaked this to make it easy to add more strategies in the future!

thomastaylor312 reviewed (changes requested):
Just a few small things. I should be able to just hit the LGTM button as soon as we get tests passing!

cohix committed:
address review feedback, add possibility of multiple verification strategies

cohix commented:
@thomastaylor312 @radu-matei this should be good to go! Hopefully tests pass, all comments addressed.

thomastaylor312 reviewed (approved):
This is looking great! Thanks for iterating over this with us and fixing tests

thomastaylor312 commented:
Ugh...looks like it doesn't pick up Go 1.16 by default. Probably easiest to use https://github.com/actions/setup-go to select the right version

cohix committed:
add setup-go step in CI

cohix commented:
@thomastaylor312 ah dang, I was wondering what was configuring that :)

Hopefully good now?

radu-matei reviewed (approved):
LGTM (once CI is green).

Thanks a lot for your work, @cohix!

radu-matei commented:
Final thing, @cohix  — I don't want to squash all your commits, would you please mind rebasing and keeping the useful ones?

Thank you so much!

cohix commented:
@radu-matei squashed the useless ones :) CI is green so merge at will!

cohix commented:
Thanks folks <3
